openhcl/underhill_core/src/dma_client.rs

```diff
@@ -0,0 +1,89 @@
+use std::sync::{Arc, Weak};
+use crate::dma_manager::GlobalDmaManager;
+use user_driver::dma;
+
+// DMA Client structure, representing a specific client instance
+pub struct DmaClient {
+    manager: Weak<GlobalDmaManager>,
+}
+
+impl DmaClient {
+    pub fn new(manager: Weak<GlobalDmaManager>) -> Self {
+        Self { manager }
+    }
+
+    fn pin_memory(&self, range: &MemoryRange) -> Result<usize, DmaError> {
+        let manager = self.manager.upgrade().ok_or(DmaError::InitializationFailed)?;
+        let threshold = manager.get_client_threshold(self).ok_or(DmaError::InitializationFailed)?;
+
+        if range.size <= threshold && manager.is_pinned(range) {
+            Ok(range.start)
+        } else {
+            Err(DmaError::PinFailed)
+        }
+    }
+
+    pub fn map_dma_ranges(
```

Comment:
I understand this is a draft ... could you add a doc comment so that folks know the intended contract and usage here?